src/components/authorize/jwt-service.ts

```diff
@@ -29,13 +31,65 @@ export class JwtService implements JwtServiceInterface {
     }
   }
 
-  getPayload(jwt: string): AuthorizeRequestPayload {
+  async getPayloadWithSigCheck(jwt: string): Promise<any> {
     const jwtElements = jwt.split(".");
     if (jwtElements.length !== 3) {
       throw new JwtPayloadParseError("JWT was not three elements");
     }
+
+    if ((await this.signatureCheck(jwt)) === false) {
+      throw new JwtSignatureVerificationError("Jwt Signature is not valid");
+    }
     const payload = jwtElements[1];
     const buffer = Buffer.from(payload, "base64url");
+
     return JSON.parse(buffer.toString());
   }
+
+  validateClaims(claims: any): Claims {
+    this.checkClaimsShapeAndType(claims);
+
+    this.checkClaimsProperties(claims);
+
+    return claims;
+  }
+
+  checkClaimsShapeAndType(claims: any): void {
+    const errors: string[] = [];
+
+    Object.entries(getClaimsObject()).forEach(([claim, claimValue]) => {
+      if (!Object.prototype.hasOwnProperty.call(claims, claim)) {
+        errors.push(`${claim} claim missing`);
+      } else if (typeof claims[claim] !== typeof claimValue) {
+        errors.push(`${claim} claim type is incorrect`);
+      }
+    });
+
+    if (errors.length > 0) {
+      throw new ClaimsError(errors.join("\r\n"));
+    }
+  }
+
+  checkClaimsProperties(claims: any): void {
+    const errors: string[] = [];
+    const requiredclaims = getKnownClaims();
+
+    Object.keys(requiredclaims).forEach((claim) => {
+      if (requiredclaims[claim] !== claims[claim]) {
+        errors.push(`${claim} has incorrect value`);
+      }
+    });
+
+    if (claims.exp <= Math.floor(new Date().getTime() / 1000)) {
```

Comment:
We're doing manual checks for expired (`exp`) and not yet valid (`nbf`) here - does the underlying JOSE library we're using have the capability to do these checks for us?

Reply:
We don't currently use the JOSE library in the frontend - the check here seems too small to bring in a library to do ?

Reply:
Not just for checking these but for validation of JWE/S in general - I'd much rather us use a library that has thought about lots of different edge cases than rolling our own validation, if we can.

https://github.com/panva/jose is heavily used in the rest of the programme and has built-ins for JWE/S.

I appreciate that the fact we use KMS complicates things, but it's still worth checking out if we can delegate our structural validation to something else.

Reply:
Looked into using JOSE . As you mentioned there was complications with trying to use it in our decryption due to KMS but have moved to using JOSE for our validation.